apps/extension/src/background/keyring/keyring.ts

```diff
@@ -169,23 +170,54 @@ export class KeyRing {
     return sensitiveData.text;
   }
 
-  // Store validated mnemonic
-  public async storeMnemonic(
-    mnemonic: string[],
+  // Store validated mnemonic or private key
+  public async storeAccountSecret(
+    accountSecret: AccountSecret,
     alias: string
   ): Promise<AccountStore | false> {
     await this.vaultService.assertIsUnlocked();
-    const phrase = mnemonic.join(" ");
+
+    const path = { account: 0, change: 0, index: 0 };
+
     try {
-      const mnemonic = Mnemonic.from_phrase(phrase);
-      const seed = mnemonic.to_seed();
-      const { coinType } = chains[this.chainId].bip44;
-      const path = { account: 0, change: 0, index: 0 };
-      const bip44Path = makeBip44PathArray(coinType, path);
-      const hdWallet = new HDWallet(seed);
-      const key = hdWallet.derive(new Uint32Array(bip44Path));
-      const privateKeyStringPtr = key.to_hex();
-      const sk = readStringPointer(privateKeyStringPtr, this.cryptoMemory);
+      const { sk, text, accountType } = ((): {
+        sk: string,
+        text: string,
+        accountType: AccountType
+      } => {
+        switch (accountSecret.t) {
+          case "Mnemonic":
+            const phrase = accountSecret.seedPhrase.join(" ");
+            const mnemonic = Mnemonic.from_phrase(phrase);
+            const seed = mnemonic.to_seed();
+            const { coinType } = chains[this.chainId].bip44;
+            const bip44Path = makeBip44PathArray(coinType, path);
+            const hdWallet = new HDWallet(seed);
+            const key = hdWallet.derive(new Uint32Array(bip44Path));
+            const privateKeyStringPtr = key.to_hex();
+            const sk = readStringPointer(privateKeyStringPtr, this.cryptoMemory);
+
+            mnemonic.free();
+            hdWallet.free();
+            key.free();
+            privateKeyStringPtr.free();
+
+            return { sk, text: phrase, accountType: AccountType.Mnemonic };
+
+          case "PrivateKey":
+            const { privateKey } = accountSecret;
+
+            return {
+              sk: privateKey,
+              text: privateKey,
+              accountType: AccountType.PrivateKey
```

Comment:
Yeah this makes sense to me! That tells us what we need to know if we are to sign with this account.